dvc/objects.py

```diff
@@ -0,0 +1,142 @@
+import logging
+
+from .exceptions import DvcException
+
+logger = logging.getLogger(__name__)
+
+
+class ObjectError(DvcException):
+    pass
+
+
+class ObjectFormatError(ObjectError):
+    pass
+
+
+class HashFile:
+    def __init__(self, path_info, tree, hash_info):
+        self.path_info = path_info
+        self.tree = tree
+        self.hash_info = hash_info
+
+    def __len__(self):
+        return 1
+
+    def __str__(self):
+        return f"object {self.hash_info}"
+
+    def __bool__(self):
+        return bool(self.hash_info)
+
+    def check(self):
+        actual = self.tree.get_hash(self.path_info)
+
+        logger.trace(
+            "cache '%s' expected '%s' actual '%s'",
+            self.path_info,
+            self.hash_info,
+            actual,
+        )
+
+        assert actual.name == self.hash_info.name
+        if actual.value.split(".")[0] != self.hash_info.value.split(".")[0]:
+            raise ObjectFormatError(f"{self} is corrupted")
+
+    @classmethod
+    def load(cls, odb, hash_info):
+        return odb.get(hash_info)
+
+    def commit(self, odb, **kwargs):
+        odb.add(self, **kwargs)
+
+
+class File(HashFile):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.src = None
+
+    @classmethod
+    def stage(cls, odb, path_info, tree, **kwargs):
+        hash_info = tree.get_hash(path_info, **kwargs)
+        raw = odb.get(hash_info)
+        obj = cls(raw.path_info, raw.tree, hash_info)
+        obj.src = HashFile(path_info, tree, hash_info)
+        return obj
```

Comment:
Git staging would've read the object into memory, but we obviously can't do that, so we do it the lazy way: we memorize the original file location (along with it's filesystem aka tree) so that we can later save it.